tests/test_oneof.py

```diff
@@ -0,0 +1,213 @@
+from pure_protobuf.types import int32
+
+
+@dataclasses.dataclass(frozen=True)
+class ComplexObj:
+    a: int
+    b: float
+    c: complex
+    d: bool
+
+
+def create_one_of_pair(scheme_: Tuple[OneOfPartInfo, ...], field_name: str,
+                       value1: Any, value2: Any) -> Tuple[OneOf_, OneOf_]:
+    oneof_value1 = OneOf_(scheme_)
+    oneof_value2 = OneOf_(scheme_)
+
+    setattr(oneof_value1, field_name, value1)
+    setattr(oneof_value2, field_name, value2)
+    return oneof_value1, oneof_value2
+
+
+@mark.parametrize('field_name, value1, value2', [
+    ('abc', "foo", "foo"),
+    ('foo', 5, 5),
+    ('f1', ComplexObj(1, 1., 1j, True), ComplexObj(1, 1., 1j, True))
+])
+def test_one_of_field_eq_hash(field_name: str, value1: Any, value2: Any):
+    parts = (
+        OneOfPartInfo('abc', str, 5),
+        OneOfPartInfo('foo', int, 10),
+        OneOfPartInfo('f1', ComplexObj, 11)
+    )
+    oneof_value1, oneof_value2 = create_one_of_pair(parts, field_name, value1, value2)
+
+    assert oneof_value1 == oneof_value2
+    assert hash(oneof_value1) == hash(oneof_value2)
+
+
+@mark.parametrize('field_name, value1, value2', [
+    ('abc', "foo", "abudu"),
+    ('foo', 5, -1),
+    ('f1', ComplexObj(1, 1., 1j, True), ComplexObj(2, 2., 2j, False))
+])
+def test_one_of_field_not_eq_hash(field_name: str, value1: Any, value2: Any):
+    parts = (
+        OneOfPartInfo('abc', str, 5),
+        OneOfPartInfo('foo', int, 10),
+        OneOfPartInfo('f1', ComplexObj, 10)
+    )
+    oneof_value1, oneof_value2 = create_one_of_pair(parts, field_name, value1, value2)
+
+    assert oneof_value1 != oneof_value2
+    assert hash(oneof_value1) != hash(oneof_value2)
+
+
+def test_hash_doesnt_work_with_unhashable_type():
+    @dataclasses.dataclass
+    class ComplexObj:
+        a: int
+        b: float
+
+    with raises(TypeError):
+        one_of_value = OneOf_((OneOfPartInfo('complex', ComplexObj, 1), ))
+        one_of_value.complex = ComplexObj(a=5, b=1.)
+        hash(one_of_value)
+
+
+def test_which_one_of():
+    parts = (
+        OneOfPartInfo('abc', str, 5),
+        OneOfPartInfo('foo', int, 10),
+        OneOfPartInfo('f1', ComplexObj, 10),
+        OneOfPartInfo('f2', List[int], 9)
+    )
+    values = (
+        "foo",
+        5,
+        ComplexObj(1, 1., 1j, False),
+        [1, 2, 3, 4, 5]
+    )
+
+    oneof_msg = OneOf_(parts)
+
+    for part_, val in zip(parts, values):
+        setattr(oneof_msg, part_.name, val)
+
+        assert oneof_msg.which_one_of == part_.name
+        assert getattr(oneof_msg, oneof_msg.which_one_of) == val
+
+
+def name_and_value(oneof_msg: OneOf_) -> Tuple[str, Any]:
+    return oneof_msg.which_one_of, getattr(oneof_msg, oneof_msg.which_one_of)  # type: ignore
+
+
+def test_usual_message_sets_work():
+    parts = (
+        OneOfPartInfo('abc', str, 5),
+        OneOfPartInfo('foo', int, 10),
+        OneOfPartInfo('f1', ComplexObj, 10),
+        OneOfPartInfo('f2', List[int], 9)
+    )
```

Comment:
Could you move these to `pytest.mark.parametrize`?

Reply:
This are not parameters for separate measurements/runs, it is the same scheme of one `OneOf_` obj, and 

And same stuff here:
```python
oneof_msg.foo = expected  # every set actually overrides
```

This could be changed to `getattr` call, but I decided to hardcode.

If needed, it can be dropped.

Reply:
Sorry it's me, I've overlooked it